In [ ]:
 !pip install pyscf

     |████████████████████████████████| 29.7 MB 188 kB/s 


In [ ]:
from pyscf import gto, scf
from pyscf.geomopt.berny_solver import optimize
mol = gto.M(atom='H -4.4005368234 -0.1221007613 2.241525046; H -3.892960984 0.903739148 3.0964143699; O -4.6674893014 0.2942847846 3.0623807031', basis='ccpvdz')
mf = scf.RHF(mol)
mf.kernel()
mol_eq = optimize(mf)

converged SCF energy = -76.0229049375925


-76.02290493759246

### benzene 0 PBE/ccpvdz

In [ ]:
%%time
#benzene 0
from pyscf import gto, dft
mol = gto.M(atom='''C       1.382994008433  -0.246008426169 0.14919345939;
C       0.450993992214  -1.254008443711 0.415193464018;
C       -0.947006032114 -1.022008439673 0.379193463392;
C       -1.38000603965  0.294991583245  -0.1088065451;
C       -0.46000602364  1.22099159936   -0.482806551609;
C       0.947994000863  0.998991595496  -0.318806548755;
H       2.480994027541  -0.257008426361 0.338193462679;
H       0.866993999453  -2.245008460957 0.616193467516;
H       -1.733006045793 -1.854008454152 0.535193466107;
H       -2.501006059158 0.507991586952  -0.305806548528;
H       -0.771006029052 2.199991616397  -0.961806559944;
H       1.716994014245  1.731991608253  -0.616806553941
''', basis='ccpvdz')
mf = dft.RKS(mol)
mf.xc = 'PBE'
mf.kernel()

converged SCF energy = -231.93471019647
CPU times: user 25.6 s, sys: 1.38 s, total: 27 s
Wall time: 13.8 s


### benzene 0 HF,ccsd,ccsd(t)/ccpvdz

In [ ]:
from pyscf import gto, scf, cc
from pyscf.cc import ccsd_t_lambda_slow as ccsd_t_lambda
from pyscf.grad import ccsd_t as ccsd_t_grad
mol = gto.M(atom='''C       1.382994008433  -0.246008426169 0.14919345939;
C       0.450993992214  -1.254008443711 0.415193464018;
C       -0.947006032114 -1.022008439673 0.379193463392;
C       -1.38000603965  0.294991583245  -0.1088065451;
C       -0.46000602364  1.22099159936   -0.482806551609;
C       0.947994000863  0.998991595496  -0.318806548755;
H       2.480994027541  -0.257008426361 0.338193462679;
H       0.866993999453  -2.245008460957 0.616193467516;
H       -1.733006045793 -1.854008454152 0.535193466107;
H       -2.501006059158 0.507991586952  -0.305806548528;
H       -0.771006029052 2.199991616397  -0.961806559944;
H       1.716994014245  1.731991608253  -0.616806553941''', 
basis='ccpvdz')


In [ ]:
 %%time
 mf = scf.RHF(mol).run()

converged SCF energy = -230.69516846079
CPU times: user 7.43 s, sys: 719 ms, total: 8.15 s
Wall time: 4.18 s


In [ ]:
 %%time
mycc = cc.CCSD(mf).run()

E(CCSD) = -231.5408918123733  E_corr = -0.8457233515836473
CPU times: user 2min 36s, sys: 3.02 s, total: 2min 39s
Wall time: 1min 38s


In [ ]:
 %%time
et_correction = mycc.ccsd_t()

CCSD(T) correction = -0.0378880701273352
CPU times: user 4min 27s, sys: 819 ms, total: 4min 28s
Wall time: 2min 16s


### benzene 0 PBE/ccpvtz

In [ ]:
%%time
#benzene 0
from pyscf import gto, dft
mol = gto.M(atom='''C       1.382994008433  -0.246008426169 0.14919345939;
C       0.450993992214  -1.254008443711 0.415193464018;
C       -0.947006032114 -1.022008439673 0.379193463392;
C       -1.38000603965  0.294991583245  -0.1088065451;
C       -0.46000602364  1.22099159936   -0.482806551609;
C       0.947994000863  0.998991595496  -0.318806548755;
H       2.480994027541  -0.257008426361 0.338193462679;
H       0.866993999453  -2.245008460957 0.616193467516;
H       -1.733006045793 -1.854008454152 0.535193466107;
H       -2.501006059158 0.507991586952  -0.305806548528;
H       -0.771006029052 2.199991616397  -0.961806559944;
H       1.716994014245  1.731991608253  -0.616806553941
''', basis='ccpvtz')
mf = dft.RKS(mol)
mf.xc = 'PBE'
mf.kernel()

converged SCF energy = -231.995331929734
CPU times: user 8min 36s, sys: 5.19 s, total: 8min 41s
Wall time: 4min 26s


### benzene 0 HF,ccsd,ccsd(t)/ccpvtz

In [ ]:
from pyscf import gto, scf, cc
from pyscf.cc import ccsd_t_lambda_slow as ccsd_t_lambda
from pyscf.grad import ccsd_t as ccsd_t_grad
mol = gto.M(atom='''C       1.382994008433  -0.246008426169 0.14919345939;
C       0.450993992214  -1.254008443711 0.415193464018;
C       -0.947006032114 -1.022008439673 0.379193463392;
C       -1.38000603965  0.294991583245  -0.1088065451;
C       -0.46000602364  1.22099159936   -0.482806551609;
C       0.947994000863  0.998991595496  -0.318806548755;
H       2.480994027541  -0.257008426361 0.338193462679;
H       0.866993999453  -2.245008460957 0.616193467516;
H       -1.733006045793 -1.854008454152 0.535193466107;
H       -2.501006059158 0.507991586952  -0.305806548528;
H       -0.771006029052 2.199991616397  -0.961806559944;
H       1.716994014245  1.731991608253  -0.616806553941''', 
basis='ccpvtz')

In [ ]:
 %%time
 mf = scf.RHF(mol).run()

converged SCF energy = -230.748785471703
CPU times: user 8min 47s, sys: 1.59 s, total: 8min 49s
Wall time: 4min 30s


In [ ]:
 %%time
mycc = cc.CCSD(mf).run()

E(CCSD) = -231.82238787055  E_corr = -1.073602398845827
CPU times: user 1h 23min 8s, sys: 5min 27s, total: 1h 28min 35s
Wall time: 55min 57s


In [ ]:
%%time
et_correction = mycc.ccsd_t()

CCSD(T) correction = -0.0550978627359253
CPU times: user 3h 3min 22s, sys: 1min 20s, total: 3h 4min 43s
Wall time: 1h 35min 18s


### ethane 990 PBE/ccpvdz

In [ ]:
#ethane 990
from pyscf import gto, dft
mol = gto.M(atom='''C -11.090319001 -3.0762961035 -11.2557667575;
C -9.9068724062 -2.2614359174 -11.6331980834;
H -11.9928102906 -2.5333223807 -11.0936121026;
H -10.8707774653 -3.5865231235 -10.3494665677;
H -11.2644862601 -3.9065593327 -12.0146320249;
H -10.1009652377 -1.1564606044 -11.8659012248;
H -9.5059881461 -2.5797431429 -12.6216356516;
H -9.1181726529 -2.4637111156 -10.8827337756
''', basis='ccpvdz')
mf = dft.RKS(mol)
mf.xc = 'PBE'
mf.kernel()

converged SCF energy = -79.6844271138827


-79.68442711388272

### butane 0 PBE/ccpvdz

In [ ]:
#butane 0
from pyscf import gto, dft
mol = gto.M(atom='''C -1.6886832945 0.3957759236 0.1551348391;
C -0.1629189261 0.3824401125 0.1430991632;
H -2.0486505452 1.1755266129 0.8589747687;
H -2.0808434536 -0.5897703783 0.4853266103;
H -2.0808418416 0.6236529757 -0.8589664517;
H 0.2022101721 0.1742385926 1.1725579159;
H 0.202211861 1.3852719181 -0.1690872196;
C 0.3729734386 -0.6834753622 -0.8190466782;
H 0.0078425448 -1.6863071596 -0.5068604008;
C 1.8987378076 -0.6968113395 -0.8310821516;
H 0.0078444519 -0.475273742 -1.8485054528;
H 2.2908961593 -0.92468861 0.183019168;
H 2.2908981502 0.288734964 -1.1612736941;
H 2.2587050716 -1.4765619732 -1.5349221356''', basis='ccpvdz')
mf = dft.RKS(mol)
mf.xc = 'PBE'
mf.kernel()

converged SCF energy = -158.213511336515


-158.21351133651467

### butane 0 PBE/ccpvtz

In [ ]:
#butane 0
from pyscf import gto, dft
mol = gto.M(atom='''C -1.6886832945 0.3957759236 0.1551348391;
C -0.1629189261 0.3824401125 0.1430991632;
H -2.0486505452 1.1755266129 0.8589747687;
H -2.0808434536 -0.5897703783 0.4853266103;
H -2.0808418416 0.6236529757 -0.8589664517;
H 0.2022101721 0.1742385926 1.1725579159;
H 0.202211861 1.3852719181 -0.1690872196;
C 0.3729734386 -0.6834753622 -0.8190466782;
H 0.0078425448 -1.6863071596 -0.5068604008;
C 1.8987378076 -0.6968113395 -0.8310821516;
H 0.0078444519 -0.475273742 -1.8485054528;
H 2.2908961593 -0.92468861 0.183019168;
H 2.2908981502 0.288734964 -1.1612736941;
H 2.2587050716 -1.4765619732 -1.5349221356''', basis='ccpvtz')
mf = dft.RKS(mol)
mf.xc = 'PBE'
mf.kernel()

converged SCF energy = -158.26831725028


-158.26831725028046

### butane 0 CCSDT/ccpvtz

In [ ]:
from pyscf import gto, scf, cc
from pyscf.cc import ccsd_t_lambda_slow as ccsd_t_lambda
from pyscf.grad import ccsd_t as ccsd_t_grad
mol = gto.M(atom='''C -1.6886832945 0.3957759236 0.1551348391;
C -0.1629189261 0.3824401125 0.1430991632;
H -2.0486505452 1.1755266129 0.8589747687;
H -2.0808434536 -0.5897703783 0.4853266103;
H -2.0808418416 0.6236529757 -0.8589664517;
H 0.2022101721 0.1742385926 1.1725579159;
H 0.202211861 1.3852719181 -0.1690872196;
C 0.3729734386 -0.6834753622 -0.8190466782;
H 0.0078425448 -1.6863071596 -0.5068604008;
C 1.8987378076 -0.6968113395 -0.8310821516;
H 0.0078444519 -0.475273742 -1.8485054528;
H 2.2908961593 -0.92468861 0.183019168;
H 2.2908981502 0.288734964 -1.1612736941;
H 2.2587050716 -1.4765619732 -1.5349221356''', basis='ccpvtz')
# mf = scf.RHF(mol).run()
# mycc = cc.CCSD(mf)
# mycc.kernel()


mf = scf.RHF(mol)
cc_scan = cc.CCSD(mf).as_scanner()

def f(mol):
    # Compute CCSD(T) energy
    mf = scf.RHF(mol).run()
    mycc = cc.CCSD(mf).run()
    et_correction = mycc.ccsd_t()
    e_tot = mycc.e_tot + et_correction
    return e_tot

f(mol)

converged SCF energy = -157.350305890794
E(CCSD) = -158.187172126961  E_corr = -0.8368662361683308


In [ ]:
 mf = scf.RHF(mol).run()

converged SCF energy = -157.350305890794


In [ ]:
mycc = cc.CCSD(mf).run()

E(CCSD) = -158.1871721269613  E_corr = -0.8368662361683357


In [ ]:
et_correction = mycc.ccsd_t()

CCSD(T) correction = -0.0297499881941838


In [ ]:
mycc.e_tot + et_correction

-158.21692211515546

## propane 109 PBE/ccpvdz

In [ ]:
#propane 109
from pyscf import gto, dft
mol = gto.M(atom='''C 1.2599174208 -1.199543421 1.4135412586
C 2.1303738699 -0.5773675516 0.2146135146
H 0.8679811917 -0.4187060781 2.0473868595
H 1.7438790182 -1.9058877415 1.9601275872
H 0.3011863 -1.6197072488 1.0580901381
H 2.9096888433 0.0927168972 0.5959432487
H 1.4719160163 0.103216073 -0.4280875696
C 2.7291352125 -1.5952505146 -0.7534454244
H 3.5417458323 -2.1696461192 -0.1701726671
H 3.2308218414 -1.0497941176 -1.5298452228
H 2.0049924947 -2.2991182961 -1.2541881159''', basis='ccpvdz')
for i in range(10):
    mf = dft.RKS(mol)
    mf.xc = 'PBE'
    mf.grids.level = i
    print(f"grid={i}")
    mf.kernel()

grid=0
converged SCF energy = -118.917312704806
grid=1
converged SCF energy = -118.940208284653
grid=2
converged SCF energy = -118.940210737454
grid=3
converged SCF energy = -118.940210536203
grid=4
converged SCF energy = -118.94020961001
grid=5
converged SCF energy = -118.940209518086
grid=6
converged SCF energy = -118.940209465245
grid=7
converged SCF energy = -118.940209480009
grid=8
converged SCF energy = -118.940209479043
grid=9
converged SCF energy = -118.940209511699


In [ ]:
#propane 109
from pyscf import gto, dft
mol = gto.M(atom='''C 1.2599174208 -1.199543421 1.4135412586
C 2.1303738699 -0.5773675516 0.2146135146
H 0.8679811917 -0.4187060781 2.0473868595
H 1.7438790182 -1.9058877415 1.9601275872
H 0.3011863 -1.6197072488 1.0580901381
H 2.9096888433 0.0927168972 0.5959432487
H 1.4719160163 0.103216073 -0.4280875696
C 2.7291352125 -1.5952505146 -0.7534454244
H 3.5417458323 -2.1696461192 -0.1701726671
H 3.2308218414 -1.0497941176 -1.5298452228
H 2.0049924947 -2.2991182961 -1.2541881159''', basis='ccpvtz')
for i in range(5):
    mf = dft.RKS(mol)
    mf.xc = 'PBE'
    mf.grids.level = i
    print(f"grid={i}")
    mf.kernel()

grid=0
converged SCF energy = -118.959711256604
grid=1
converged SCF energy = -118.984599043447
grid=2
converged SCF energy = -118.984619900943
grid=3
converged SCF energy = -118.984619360017
grid=4
converged SCF energy = -118.98462065184


## water ccpvtz

In [ ]:
%%time
#water 0000
from pyscf import gto, dft
mol = gto.M(atom='''O -4.6674893014 0.2942847846 3.0623807031
H -3.892960984 0.903739148 3.0964143699
H -4.4005368234 -0.1221007613 2.241525046
''', basis='ccpvtz')
mf = dft.RKS(mol)
mf.xc = 'PBE'
mf.kernel()

converged SCF energy = -76.3707642911744
CPU times: user 2.03 s, sys: 842 ms, total: 2.87 s
Wall time: 1.66 s


In [ ]:
from pyscf import gto, scf, cc
from pyscf.cc import ccsd_t_lambda_slow as ccsd_t_lambda
from pyscf.grad import ccsd_t as ccsd_t_grad
mol = gto.M(atom='''O -4.6674893014 0.2942847846 3.0623807031
H -3.892960984 0.903739148 3.0964143699
H -4.4005368234 -0.1221007613 2.241525046
''', basis='ccpvtz')

In [ ]:
%%time
mf = scf.RHF(mol).run()
mycc = cc.CCSD(mf).run()
et_correction = mycc.ccsd_t()
print(f"E(CCSDT)={mycc.e_tot + et_correction}")

converged SCF energy = -76.052432958174
E(CCSD) = -76.3351095988434  E_corr = -0.2826766406694271
CCSD(T) correction = -0.00802736925395756
E(CCSDT)=-76.34313696809735
CPU times: user 3.43 s, sys: 1.07 s, total: 4.51 s
Wall time: 2.58 s


In [ ]:
mycc = cc.CCSD(mf).run()

E(CCSD) = -76.33510959884327  E_corr = -0.2826766406694274


In [ ]:
et_correction = mycc.ccsd_t()

CCSD(T) correction = -0.00802736925395758


In [ ]:
mycc.e_tot + et_correction

-76.34313696809723

| Water 0000 | HF | CCSD | CCSDT | DFT(PBE)|
| --- | --------- |----|-------|----|
| MOLPRO(cheng dataset)@ccpvtz| -76.052432958180|-76.321776337147 | -76.329392417032 |N/A|
| psi4@ccpvtz       |-76.052432958130 |-76.335109594845|-76.343136964956|-78.510651785425|
|pyscf@ccpvtz        | -76.052432958173|-76.335109598843|-76.343136968097|-76.370764291174

## water CCSDT/ccpvdz

In [ ]:
from pyscf import gto, scf, cc
from pyscf.cc import ccsd_t_lambda_slow as ccsd_t_lambda
from pyscf.grad import ccsd_t as ccsd_t_grad
from pyscf.geomopt import berny_solver

mol = gto.M(
    verbose = 3,
    atom = [
        ['O' , (0. , 0.     , 0.    )],
        ['H' , (0. ,-0.757  ,-0.587)],
        ['H' , (0. , 0.757  ,-0.587)]],
    basis = 'ccpvdz'
)
mf = scf.RHF(mol)
cc_scan = cc.CCSD(mf).as_scanner()

def f(mol):
    # Compute CCSD(T) energy
    mf = scf.RHF(mol).run()
    mycc = cc.CCSD(mf).run()
    et_correction = mycc.ccsd_t()
    e_tot = mycc.e_tot + et_correction

    # Compute CCSD(T) gradients
    g = ccsd_t_grad.Gradients(mycc).kernel()
    print('CCSD(T) nuclear gradients:')
    print(g)
    return e_tot, g

fake_method = berny_solver.as_pyscf_method(mol, f)

new_mol = berny_solver.optimize(fake_method)

print('Old geometry (Bohr)')
print(mol.atom_coords())

print('New geometry (Bohr)')
print(new_mol.atom_coords())


Geometry optimization cycle 1
Cartesian coordinates (Angstrom)
 Atom        New coordinates             dX        dY        dZ
   O   0.000000   0.000000   0.000000    0.000000  0.000000  0.000000
   H   0.000000  -0.757000  -0.587000    0.000000  0.000000  0.000000
   H   0.000000   0.757000  -0.587000    0.000000  0.000000  0.000000
converged SCF energy = -76.0267656731181
E(CCSD) = -76.24010891963178  E_corr = -0.2133432465136902
CCSD(T) correction = -0.00306002383685392


/usr/local/lib/python3.7/dist-packages/pyscf/cc/ccsd_rdm.py:52: H5pyDeprecationWarning: dataset.value has been deprecated. Use dataset[()] instead.
  d2 = (f['dovov'].value, f['dvvvv'].value, f['doooo'].value, f['doovv'].value,
/usr/local/lib/python3.7/dist-packages/pyscf/cc/ccsd_rdm.py:53: H5pyDeprecationWarning: dataset.value has been deprecated. Use dataset[()] instead.
  f['dovvo'].value, None,             f['dovvv'].value, f['dooov'].value)


--------------- CCSD gradients ---------------
         x                y                z
0 O     0.0000000000     0.0000000000    -0.0121989309
1 H    -0.0000000000     0.0024680759     0.0060994655
2 H    -0.0000000000    -0.0024680759     0.0060994655
----------------------------------------------
CCSD(T) nuclear gradients:
[[ 3.28303019e-16  5.85264478e-16 -1.21989309e-02]
 [-1.64503467e-18  2.46807595e-03  6.09946546e-03]
 [-3.26657985e-16 -2.46807595e-03  6.09946546e-03]]
cycle 1: E = -76.2431689435  dE = -76.2432  norm(grad) = 0.0153429

Geometry optimization cycle 2
Cartesian coordinates (Angstrom)
 Atom        New coordinates             dX        dY        dZ
   O   0.000000  -0.000000   0.010029    0.000000 -0.000000  0.010029
   H   0.000000  -0.752571  -0.592014    0.000000  0.004429 -0.005014
   H   0.000000   0.752571  -0.592014    0.000000 -0.004429 -0.005014
converged SCF energy = -76.0263690245671
E(CCSD) = -76.24028061063756  E_corr = -0.2139115860704651
CCSD(T) co

In [ ]:
pip install -U pyberny

In [ ]:
import numpy as np

In [ ]:
np.linalg.norm(new_mol.atom_coords()[1]-new_mol.atom_coords()[0])

1.8227625873740765

In [ ]:
new_mol.atom_coords()

array([[ 0.00000000e+00,  4.99554274e-16,  2.44048971e-02],
       [ 0.00000000e+00, -1.41754376e+00, -1.12147168e+00],
       [ 0.00000000e+00,  1.41754376e+00, -1.12147168e+00]])

In [ ]:
OH1 = new_mol.atom_coords()[1]-new_mol.atom_coords()[0]
OH2 = new_mol.atom_coords()[2]-new_mol.atom_coords()[0]
np.arccos(np.dot(OH1, OH2)/(np.linalg.norm(OH1)*np.linalg.norm(OH2)))/np.pi*180

102.0990680685159